presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java

```diff
@@ -195,14 +195,65 @@ public static Slice reverse(@SqlType("varchar(x)") Slice slice)
     @SqlType(StandardTypes.BIGINT)
     public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
     {
+        return stringPositionUtil(string, substring, 1, false);
+    }
+
+    @Description("returns index of n-th occurrence of a substring (or 0 if not found)")
+    @ScalarFunction("strpos")
+    @LiteralParameters({"x", "y"})
+    @SqlType(StandardTypes.BIGINT)
+    public static long stringPosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring, @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return stringPositionUtil(string, substring, instance, false);
+    }
+
+    @Description("returns index of first occurrence of a substring from the back (or 0 if not found)")
+    @ScalarFunction("strrpos")
+    @LiteralParameters({"x", "y"})
+    @SqlType(StandardTypes.BIGINT)
+    public static long stringReversePosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring)
+    {
+        return stringPositionUtil(string, substring, 1, true);
+    }
+
+    @Description("returns index of n-th occurrence of a substring from the back (or 0 if not found)")
+    @ScalarFunction("strrpos")
+    @LiteralParameters({"x", "y"})
+    @SqlType(StandardTypes.BIGINT)
+    public static long stringReversePosition(@SqlType("varchar(x)") Slice string, @SqlType("varchar(y)") Slice substring, @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return stringPositionUtil(string, substring, instance, true);
+    }
+
+    private static long stringPositionUtil(Slice string, Slice substring, long instance, boolean countBackward)
+    {
+        if (instance <= 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid instance argument");
+        }
         if (substring.length() == 0) {
             return 1;
         }
 
-        int index = string.indexOf(substring);
-        if (index < 0) {
-            return 0;
+        int foundInstances = 0;
+        // set the initial index beyond the either ends of the string
+        // this is to allow for the initial index decrement/increment
+        int index = countBackward ? string.length() : -1;
```

Comment:
It makes more sense to have two separate functions to handle counting forward vs counting backward than putting them into one function. It would be easier to read, and the function would not have as many branches. Do you mind separate them into two functions? Thanks!

Reply:
I have added two different methods `stringPositionFromStart(string, substring, instance)` and `stringPositionFromEnd(string, substring, instance)`